Add a manage clusters page to the multi-cluster nav to list/connect/disconnect/edit clusters in multi-cluster setup

aaSharma14 opened the pull request:
Add a manage clusters page to the multi-cluster nav to list/connect/disconnect/edit clusters in multi-cluster setup


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts:
```diff
@@ -35,14 +45,61 @@ export class MultiClusterFormComponent implements OnInit, OnDestroy {
+      this.remoteClusterForm.get('clusterFsid').setValue(this.cluster.name);
+      this.remoteClusterForm.get('clusterFsid').disable();
+    }
+    this.clusterAliasNames = this.clustersData.map((cluster) => {
+      return cluster['cluster_alias'];
+    });
+
+    this.clusterUrls = this.clustersData.map((cluster) => {
+      return cluster['url'];
+    });
+
+    this.clusterUsers = this.clustersData.map((cluster) => {
+      return cluster['user'];
+    });
```
maybe something like this should work?
```suggestion
    [this.clusterAliasNames, this.clusterUrls, this.clusterUsers] = ['cluster_alias', 'url', 'user'].map(prop => this.clustersData?.map(cluster => cluster[prop]));
```

nizamial09 reviewed (changes requested):
initial review before proceeding further.

The Connect Cluster form in the Overview page is broken for me. 

![Screenshot from 2024-02-02 10-09-08](https://github.com/ceph/ceph/assets/71764184/9d9e83e6-863d-4b44-886d-2e7d881b2367)

After clicking the button it just stays with this loading spinner
![image](https://github.com/ceph/ceph/assets/71764184/1e875cbf-5b44-487b-8749-498ccf7731b7)


Notice the button name `null Cluster`
![Screenshot from 2024-02-02 10-12-38](https://github.com/ceph/ceph/assets/71764184/fadf3253-4ac7-4b8d-8716-99105b415a06)

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/api/multi-cluster.service.ts:
```diff
@@ -18,6 +20,44 @@ export class MultiClusterService {
       .subscribe(this.getClusterObserver());
   }
 
+  startClusterTokenStatusPolling() {
+    let clustersTokenMap = new Map<string, { token: string; user: string }>();
+    const dataSubscription = this.subscribe((resp: any) => {
+      const clustersConfig = resp['config'];
+      const tempMap = new Map<string, { token: string; user: string }>();
+      if (clustersConfig) {
+        Object.keys(clustersConfig).forEach((clusterKey: string) => {
+          const clusterDetailsList = clustersConfig[clusterKey];
+          clusterDetailsList.forEach((clusterDetails: any) => {
+            if (clusterDetails['token'] && clusterDetails['name'] && clusterDetails['user']) {
+              tempMap.set(clusterDetails['name'], {
+                token: clusterDetails['token'],
+                user: clusterDetails['user']
+              });
+            }
+          });
+        });
+
+        if (tempMap.size > 0) {
+          clustersTokenMap = tempMap;
+          dataSubscription.unsubscribe();
+          this.checkAndStartTimer(clustersTokenMap);
+        }
+      }
+    });
+  }
```
i did not completely understood this, could you please briefly explain?

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/api/multi-cluster.service.ts:
```diff
@@ -18,6 +20,44 @@ export class MultiClusterService {
+  private checkAndStartTimer(clustersTokenMap: Map<string, { token: string; user: string }>) {
+    this.checkTokenStatus(clustersTokenMap).subscribe(this.getClusterTokenStatusObserver());
+    this.timerService
+      .get(() => this.checkTokenStatus(clustersTokenMap), 600000)
```
its better to initialize a constant like this on top of the class and reuse it below
`TOKEN_CHECK_INTERVAL = 600000 // 10m interval`

```suggestion
      .get(() => this.checkTokenStatus(clustersTokenMap), TOKEN_CHECK_INTERVAL)
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/app-routing.module.ts:
```diff
@@ -189,6 +190,13 @@ const routes: Routes = [
         path: 'multi-cluster',
         component: MultiClusterComponent
       },
+      {
+        path: 'manage-clusters',
+        component: MultiClusterListComponent,
+        data: {
+          breadcrumbs: 'Multi-Cluster/Manage Clusters'
+        }
+      },
```
this should be a child component of the `multi-cluster` path, not a separate entity

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.ts:
```diff
@@ -35,14 +45,61 @@ export class MultiClusterFormComponent implements OnInit, OnDestroy {
+          } else if (this.remoteClusterForm) {
+            remoteClusterUrl = this.remoteClusterForm.getValue('remoteClusterUrl');
+          }
+          return (
+            this.remoteClusterForm &&
+            this.clusterUrls.includes(remoteClusterUrl) &&
+            this.clusterUsers.includes(username)
```
this will break if the clusterData is empty

```suggestion
          return (
            this.remoteClusterForm &&
            this.clusterUrls?.includes(remoteClusterUrl) &&
            this.clusterUsers?.includes(username)
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts:
```diff
@@ -76,7 +76,11 @@ export class ApiInterceptorService implements HttpInterceptor {
-    const apiUrl = localStorage.getItem('cluster_api_url');
+    let apiUrl = localStorage.getItem('cluster_api_url');
+
+    if (apiUrl && !apiUrl.endsWith('/')) {
+      apiUrl += '/';
+    }
```
this is being repeated in the python as well as now in front-end. Do we need to duplicate them?

nizamial09 reviewed (changes requested):
I didn't deep-dived into the code and more focused on the functionality but there are some doubts I have. PTAL.

the order should be
Connect, Edit, Reconnect and Disconnect
![image](https://github.com/ceph/ceph/assets/71764184/8da59de6-9a43-4454-863f-7f7893390262)


I opened a form to `Connect` Cluster but the heading and button says `Add` Cluster. When we are renaming we should be careful to replace all the other placeholders including the toast notification we show when the action is successful.
![Screenshot from 2024-02-02 10-17-34](https://github.com/ceph/ceph/assets/71764184/47acf175-1d22-438d-a404-cfa2e4a791a6)


Editing is not working

https://github.com/ceph/ceph/assets/71764184/d271f040-c861-4aae-8fb9-8316d1cbba62

When I reconnected it says
![image](https://github.com/ceph/ceph/assets/71764184/52e9e9e8-2ec3-4096-90ba-9007f2508f72)

it should be something like `Cluster successfully reconnected` or something similar


Its forcing me to click twice on the `Auth with Token` checkbox inorder to switch to the token window

https://github.com/ceph/ceph/assets/71764184/4430f79b-0947-4f00-ac39-e4e8c314bda1


the `CHECKING` is taking some time to check even though I can switch to kcli cluster
![image](https://github.com/ceph/ceph/assets/71764184/dca22098-44d5-45f1-808d-ce2746141c74)

nizamial09 commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -38,6 +38,7 @@ class Auth(RESTController, ControllerAuthMixin):
     Provide authenticates and returns JWT token.
     """
     # pylint: disable=R0912
+
```
unintended change

nizamial09 reviewed (approved):
everything's working. 

once the cluster is disconnected, we should remove the fsid of that cluster from config.

Pegonzal commented:
Hey @aaSharma14, when I try to run a cluster to test this I'm getting:

```
2024-02-06T15:04:59.239+0000 7f399d5b3700 -1 Traceback (most recent call last):
  File "/ceph/src/pybind/mgr/dashboard/module.py", line 346, in serve
    mapper, parent_urls = Router.generate_routes(self.url_prefix)
  File "/ceph/src/pybind/mgr/dashboard/controllers/_router.py", line 49, in generate_routes
    controllers = BaseController.load_controllers()
  File "/ceph/src/pybind/mgr/dashboard/controllers/_base_controller.py", line 45, in load_controllers 
    importlib.import_module(f'{__package__}.{module}')
  File "/usr/lib64/python3.6/importlib/__init__.py", line 126, in import_module 
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/ceph/src/pybind/mgr/dashboard/controllers/multi_cluster.py", line 6, in <module>
    import jwt
ModuleNotFoundError: No module named 'jwt'
```

aaSharma14 commented:
> Hey @aaSharma14, when I try to run a cluster to test this I'm getting:
> 
> ```
> 2024-02-06T15:04:59.239+0000 7f399d5b3700 -1 Traceback (most recent call last):
>   File "/ceph/src/pybind/mgr/dashboard/module.py", line 346, in serve
>     mapper, parent_urls = Router.generate_routes(self.url_prefix)
>   File "/ceph/src/pybind/mgr/dashboard/controllers/_router.py", line 49, in generate_routes
>     controllers = BaseController.load_controllers()
>   File "/ceph/src/pybind/mgr/dashboard/controllers/_base_controller.py", line 45, in load_controllers 
>     importlib.import_module(f'{__package__}.{module}')
>   File "/usr/lib64/python3.6/importlib/__init__.py", line 126, in import_module 
>     return _bootstrap._gcd_import(name[level:], package, level)
>   File "<frozen importlib._bootstrap>", line 994, in _gcd_import
>   File "<frozen importlib._bootstrap>", line 971, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
>   File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
>   File "<frozen importlib._bootstrap_external>", line 678, in exec_module
>   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
>   File "/ceph/src/pybind/mgr/dashboard/controllers/multi_cluster.py", line 6, in <module>
>     import jwt
> ModuleNotFoundError: No module named 'jwt'
> ```

@Pegonzal , am not sure. I tried running it today on top of this branch and its working fine. Can you please retry?

nizamial09 commented:
I just remembered that @Pegonzal might be onto something here. We removed the implementation of jwt from our production code while ago from https://github.com/ceph/ceph/pull/54710/files#diff-17c3c754462ec3d2e3614f52001a501bed033a28fc0d7111c0944817b8446ffbL12. And it was for a specific reason and a wider known issue. So I don't think bringing back (or importing it) might be good again. And moreover we removed the jwt library from the `ceph.spec.in` file as well. 

@aaSharma14 So you need to reuse the newer implementation in your code to make it work and remove the jwt import from the multi_cluster.py.

aaSharma14 commented:
> I just remembered that @Pegonzal might be onto something here. We removed the implementation of jwt from our production code while ago from https://github.com/ceph/ceph/pull/54710/files#diff-17c3c754462ec3d2e3614f52001a501bed033a28fc0d7111c0944817b8446ffbL12. And it was for a specific reason and a wider known issue. So I don't think bringing back (or importing it) might be good again. And moreover we removed the jwt library from the `ceph.spec.in` file as well.
> 
> @aaSharma14 So you need to reuse the newer implementation in your code to make it work and remove the jwt import from the multi_cluster.py.

Thanks @nizamial09 for the info. @Pegonzal @nizamial09 I have updated the PR with the required changes

Pegonzal reviewed (approved):
Thanks @aaSharma14 ! It worked pretty good
